### PR TITLE
PHOENIX-1674 Snapshot isolation transaction support through Tephra

### PR DESCRIPTION
**Pull request by twdsilva**
PR that implements support for transactions using Tephra.

@JamesRTaylor  @elilevine  @jfernandosf  @samarthjain  @mujtabachohan

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexMetadataIT.java [hunk 1]
```diff
@@ -100,7 +100,7 @@ private static void assertActiveIndex(Connection conn, String schemaName, String
         String fullTableName = SchemaUtil.getTableName(schemaName, tableName);
         conn.createStatement().executeQuery("SELECT count(*) FROM " + fullTableName).next(); // client side cache will update
         PhoenixConnection pconn = conn.unwrap(PhoenixConnection.class);
-        pconn.getMetaDataCache().getTable(new PTableKey(pconn.getTenantId(), fullTableName)).getIndexMaintainers(ptr, pconn);
```
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexMetadataIT.java]: Probably good to factor out this method change (i.e. having PhoenixConnection.getTable()) into a separate pull to master as it appears to have generated a lot of diffs unrelated to transactions.

File: phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java [hunk 2]
```diff
@@ -259,10 +259,10 @@ private PhoenixConnection getConnectionWithTableOrderPreservingMutationState() t
         Connection con = driver.connect(url, new Properties());
         PhoenixConnection phxCon = new PhoenixConnection(con.unwrap(PhoenixConnection.class));
         final Map<TableRef,Map<ImmutableBytesPtr,MutationState.RowMutationState>> mutations = Maps.newTreeMap(new TableRefComparator());
-        return new PhoenixConnection(phxCon) {
+        return new PhoenixConnection(phxCon, null) {
```
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java]: Probably a good idea to still have a PhoenixConnection(PhoenixConnection conn) copy constructor that just calls the other one with a null second arg (and put that in a separate pull request).

File: phoenix-core/src/main/java/org/apache/phoenix/compile/PostDDLCompiler.java [hunk 3]
```diff
@@ -155,7 +180,14 @@ public boolean hasUDFs() {
                         };
                         PhoenixStatement statement = new PhoenixStatement(connection);
                         StatementContext context = new StatementContext(statement, resolver, scan, new SequenceManager(statement));
-                        ScanUtil.setTimeRange(scan, timestamp);
+                        long ts = timestamp;
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/PostDDLCompiler.java]: Does one of your TransactionUtil functions do the same thing as this code?

File: phoenix-core/src/main/java/org/apache/phoenix/compile/StatementPlan.java [hunk 4]
```diff
@@ -29,4 +33,6 @@
     ParameterMetaData getParameterMetaData();
     
     ExplainPlan getExplainPlan() throws SQLException;
+    public Set<TableRef> getSourceRefs();
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/StatementPlan.java]: This interface change (along with the one in MutationPlan) caused some diffs not related to transactions. If it's easy to merge that back separately, that might help reduce the extraneous, non transactional changes.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java [hunk 5]
```diff
@@ -1222,7 +1228,7 @@ public void createTable(RpcController controller, CreateTableRequest request,
                 // exists without making an additional query
                 PTable table =
                         loadTable(env, tableKey, cacheKey, clientTimeStamp, HConstants.LATEST_TIMESTAMP);
-                if (table != null) {
+                if (table != null && !isTableDeleted(table)) {
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java]: Is this a UDF bug?

File: phoenix-core/src/main/java/org/apache/phoenix/execute/HashJoinPlan.java [hunk 6]
```diff
@@ -411,7 +431,12 @@ public void postProcess(Object result, HashJoinPlan parent)
...
         }
-        
+
+
+        @Override
+        public QueryPlan getInnerPlan() {
+            return plan;
+        }
     }
 }
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/HashJoinPlan.java]: Looks like we're maxing out GitHub -- no more diffs below here. Maybe if the other refactors it would be better?

File: phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java (hunk 2 above)
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java]: Never mind on this change - passing null is actually required here, though might be worth a code comment on why (because it forces the connection.newMutationState() to be used to create the MutationState).

Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java]: I have a patch out for this one on PHOENIX-2409

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ConnectionQueryServicesTestImpl.java [hunk 7]
```diff
@@ -35,7 +35,7 @@
 public class ConnectionQueryServicesTestImpl extends ConnectionQueryServicesImpl {
     protected int NUM_SLAVES_BASE = 1; // number of slaves for the cluster
     
-    public ConnectionQueryServicesTestImpl(QueryServices services, ConnectionInfo info) throws SQLException {
+    public ConnectionQueryServicesTestImpl(QueryServices services, ConnectionInfo info, Properties info2) throws SQLException {
         super(services, info, null);
```
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/ConnectionQueryServicesTestImpl.java]: Should this be passing info2 into the super() instead of null?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java [hunk 8]
```diff
@@ -17,428 +17,92 @@
...
-import org.apache.phoenix.util.QueryUtil;
+import org.apache.phoenix.util.SchemaUtil;
+import org.apache.phoenix.util.TestUtil;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
 
 
+@RunWith(Parameterized.class)
```
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java]: If easy, can we do this test improvement in a separate change in master to reduce the noise here?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexMetadataIT.java (hunk 1 above)
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexMetadataIT.java]: FYI, I made this change to master, so if you could rebase the txn branch, these diffs should go away.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/StatementPlan.java (hunk 4 above)
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/StatementPlan.java]: Refactored in PHOENIX-2413.

File: phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java (hunk 2 above)
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/execute/PartialCommitIT.java]: I will add a comment

File: phoenix-core/src/main/java/org/apache/phoenix/compile/PostDDLCompiler.java (hunk 3 above)
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/PostDDLCompiler.java]: No, also this is the only place where TransactionUtil.convertToNanoseconds(ts) is called.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java (hunk 5 above)
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java]: I am not sure if this bug is related to UDFs. Some txn tests were failing because table.getName() was null.

**Comment by twdsilva** [comment]
Created new pull PR
https://github.com/apache/phoenix/pull/127
